Improve definition of non-existent origin

jugglinmike opened the pull request:
The first commit on this branch is intended to resolve gh-10560. The second is intended to resolve gh-10754.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/10878)
<!-- Reviewable:end -->

jugglinmike committed:
[wptserve] Correct domain configuration on *nix

Because Windows and UNIX-like systems differ in how they interpret the
IP address `0.0.0.0`, the hosts file contents generated for use with the
WPT server must take the platform into account.

Extend the hosts file generation logic and introduce "infrastructure
assumptions" tests to assert the inavailibility of the domain.

jugglinmike committed:
[infra] Rename nonexistent domain

Contributors have expressed confusion over the term "nonexistent origin"
to describe a domain that is not intended to exist. Rename the subdomain
to "nonexistent" to clarify its role.

annevk commented:
Given we use it as subdomain prefix, maybe we should just call it "nonexistent"? Domain gives the impression you can use it standalone.

jugglinmike commented:
Yeah, that's probably better. Since that's different than what @gsnedders originally agreed to, I'd like to avoid churn and get their feedback before making the change. What do you say, Geoffrey?

gsnedders commented:
Ask others too, but that sounds good to me.

jugglinmike commented:
@gsnedders Pitched the idea to @jgraham, and he was over the moon. Normally I'd make the change with a "fixup!" commit, but recent changes to `master` introduced conflicts, so I've rebased and squashed. Mind taking another look?

gsnedders commented on tools/serve/serve.py:
```diff
@@ -447,8 +448,17 @@ def make_hosts_file(config, host):
 
-    for not_domain in config.not_domains_set:
-        rv.append("0.0.0.0\t%s\n" % not_domain)
+    # Windows interpets the IP address 0.0.0.0 as "non-existent, making it an
```
Needs a closing double quote. (Also, where's the quote from? I can't find this documented anywhere?)

jugglinmike replied:
I've added the double quote in a fixup! commit. I'll respond to your question in the main thread so the conversation doesn't get lost when GitHub collapses this comment.

jugglinmike commented:
> (Also, where's the quote from? I can't find this documented anywhere?)

That's not cited because it's not a quotation. I haven't seen a particularly authoritative source describing this behavior; the explanation is a result of the independent research reported at the referenced GitHub issue and a smattering of articles/blog posts like https://www.lifewire.com/four-zero-ip-address-818384 I can add a link to that article if you like.

gsnedders commented:
@jugglinmike just the quotation marks make it look like it is a quote from somewhere; if there's nowhere authoritative I'd say just drop the quotation marks

jugglinmike commented:
Oh, now I understand. I've removed them and fixed an unrelated spelling error (from "syystem" to "system")

gsnedders commented on docs/_running-tests/chrome_android.md:
```diff
@@ -19,7 +19,7 @@ Until we find a better way, we need to root the Android device and update the
 127.0.0.1   xn--n8j6ds53lwwkrqhv28a.web-platform.test
 127.0.0.1   xn--lve-6lad.web-platform.test
-0.0.0.0     nonexistent-origin.web-platform.test
+0.0.0.0     nonexistent.web-platform.test
```
Presumably as with all other Linux systems, this should be dropped.

jugglinmike replied:
I haven't verified any of this from an Android system, so I don't feel comfortable changing the instructions. But if I really want to resolve this issue, then that is my responsibility. I'll work on running WPT there, and I'll let you know what I find.

gsnedders commented on tools/serve/test_serve.py:
```diff
@@ -8,13 +9,36 @@
-def test_make_hosts_file():
+@pytest.mark.skipif(platform.uname()[0] == "Windows",
+                    reason="Expected contents are platform-dependent")
+def test_make_hosts_file_nix():
```
I think I'd rather just have one function here, with something like:

```
expected = {"",
                          "192.168.42.42\tfoo.bar",
                          "192.168.42.42\tfoo2.bar",
                          "192.168.42.42\twww.foo.bar",
                          "192.168.42.42\twww.foo2.bar",
                          "192.168.42.42\twww1.foo.bar",
                          "192.168.42.42\twww1.foo2.bar",
                          "192.168.42.42\twww2.foo.bar",
                          "192.168.42.42\twww2.foo2.bar",
                          "192.168.42.42\txn--lve-6lad.foo.bar",
                          "192.168.42.42\txn--lve-6lad.foo2.bar",
                          "192.168.42.42\txn--n8j6ds53lwwkrqhv28a.foo.bar",
                          "192.168.42.42\txn--n8j6ds53lwwkrqhv28a.foo2.bar"}
if platform.uname()[0] != "Windows":
    expected ^= {"0.0.0.0\tnonexistent-origin.foo.bar",
                           "0.0.0.0\tnonexistent-origin.foo2.bar"}
```

(except, like, don't have the whitespace a mess from copy/pasting stuff in GitHub :))

jugglinmike replied:
I considered that, but I chose this organization to simplify the test's semantics. My preference is for tests' expectations to be static so that passing/failing always means the same thing. This has introduced some duplication, though. Is that why you're recommending the tests be consolidated?  Would you be satisfied if we kept both tests but defined the second dict in terms of the first *outside* of the test bodies?

jugglinmike commented:
Following a failed attempt to root an Android device, I've asked @Hexcles if
he could verify this behavior for me. Here are the steps we're using to check
the behavior:

1. Add the following entry to the `/etc/hosts` file on the Android device:

       0.0.0.0 localhost
2. Run the following command from your development machine:

       adb reverse tcp:8000 tcp:8000
3. Run a web server bound to port 8000 on your development machine, e.g.

       python -m SimpleHTTPServer 8000
4. Visit http://0.0.0.0:8000 from the Android device

We're curious to know if that resolves to the web server on your development machine or if it fails to resolve at all

jugglinmike commented:
Speaking on IRC, @hexcles reported:

> (01:44:49 PM) Hexcles: Tl;dr. Chrome on Android behaves the same as Linux
> regarding 0.0.0.0.

So that settles it. I've updated the documentation for Android devices by removing that entry from the recommended hosts file modification. Since this change needs to happen in the first commit but interferes with the second, I've rebased this branch to make merging easier. The original version of the branch is available here:

https://github.com/w3c/web-platform-tests/compare/master...bocoup:non-existent-orig

@gsnedders would you mind taking another look?

jugglinmike commented:
Thanks, @gsnedders!
